File: deno/day07.ts
import { DaySolution, parseData } from './_tools.ts';
import { parseLines } from './_helpers.ts';

type Problem = {
  target: number;
  numbers: number[];
};

type Data = Problem[];

enum Op {
  ADD,
  MUL,
  CONCAT,
}
const OpsP1 = [Op.ADD, Op.MUL];
const OpsP2 = [Op.ADD, Op.MUL, Op.CONCAT];

function applyOp(op: Op, a: number, b: number): number {
  switch (op) {
    case Op.ADD:
      return a + b;
    case Op.MUL:
      return a * b;
    case Op.CONCAT:
      return parseInt(a.toString() + b.toString());
  }
}

function checkRecOps(numbers: number[], target: number, baseOps: Op[], previousRes: number, idx: number): number {
  if (idx === numbers.length) {
    return previousRes;
  }

  for (const op of baseOps) {
    const res = checkRecOps(numbers, target, baseOps, applyOp(op, previousRes, numbers[idx]), idx + 1);
    if (res === target) {
      return res;
    }
  }

  return 0;
}

export class Day07 implements DaySolution<Data, number> {
  day = 7;

  parseData(dataLabel: 'Example' | 'Input'): Data {
    return parseLines(parseData(this.day, dataLabel), (line) => {
      const [targetS, numbersS] = line.split(':').map((s) => s.trim());

      return {
        target: parseInt(targetS),
        numbers: numbersS.split(' ').map((s) => parseInt(s.trim())),
      };
    });
  }

  part1(data: Data): number {
    return data.map(({target, numbers}) => checkRecOps(numbers, target, OpsP1, numbers[0], 1))
      .reduce((sum, res) => sum + res, 0);
  }

  part2(data: Data): number {
    return data.map(({target, numbers}) => checkRecOps(numbers, target, OpsP2, numbers[0], 1))
      .reduce((sum, res) => sum + res, 0);
  }
}

File: rust/src/day07.rs
use crate::tools::read_raw_data;
use crate::DaySolution;

#[derive(Clone, Debug)]
enum Op {
    Add,
    Mul,
    Concat,
}

impl Op {
    fn apply(&self, a: i128, b: i128) -> i128 {
        match self {
            Op::Add => a + b,
            Op::Mul => a * b,
            Op::Concat => format!("{}{}", a, b).parse().unwrap()
        }
    }
}

const OPS_1: [Op; 2] = [Op::Add, Op::Mul];
const OPS_2: [Op; 3] = [Op::Add, Op::Mul, Op::Concat];

#[derive(Debug)]
pub struct Problem(i128, Vec<i128>);
impl From<&str> for Problem {
    fn from(s: &str) -> Self {
        let [target_s, numbers_s] = s.split(':').collect::<Vec<&str>>()[..] else { panic!("Invalid input") };

        Problem(
            target_s.parse().unwrap(),
            numbers_s.trim().split(' ').map(|n| n.parse().unwrap()).collect(),
        )
    }
}
impl Problem {
    fn check(&self, base_opts: Vec<Op>) -> bool {
        self.check_rec(base_opts, self.1[0], 1).is_some()
    }

    fn check_rec(&self, base_opts: Vec<Op>, previous_res: i128, idx: usize) -> Option<i128> {
        if idx == self.1.len() {
            if previous_res == self.0 {
                Some(previous_res)
            } else {
                None
            }
        } else {
            base_opts.iter()
                .filter_map(|op| {
                    self.check_rec(base_opts.clone(), op.apply(previous_res, self.1[idx]), idx + 1)
                })
                .next()
        }
    }
}

pub type Data = Vec<Problem>;

pub struct Day07;

impl DaySolution<Data, i128> for Day07 {
    fn new() -> Self {
        Self {}
    }

    fn day(&self) -> u16 {
        7
    }

    fn parse_input(&mut self, example: bool) -> Data {
        let raw_data = read_raw_data(7, example);
        raw_data.lines().map(|l| l.into()).collect()
    }

    fn part1(&mut self, input: &Data) -> i128 {
        input
            .iter()
            .filter(|p| p.check(OPS_1.to_vec()))
            .map(|problem| problem.0)
            .sum()
    }

    fn part2(&mut self, input: &Data) -> i128 {
        input
            .iter()
            .filter(|p| p.check(OPS_2.to_vec()))
            .map(|problem| problem.0)
            .sum()
    }
}
